compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -985,7 +985,180 @@ OMR::Power::TreeEvaluator::vstoreiEvaluator(TR::Node *node, TR::CodeGenerator *c
+
+   //Copy result into GPR
+   if (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P9))
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrld, node, resReg, tempRes);
+   else
+   {
+      generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::xxpermdi, node, tempRes, tempRes, temp, 3); //move sum to upper doubleword element of tempRes
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrd, node, resReg, tempRes);
+   }
+
+   cg->stopUsingRegister(tempRes);
+   cg->stopUsingRegister(temp);
+   cg->decReferenceCount(firstChild);
+
+   return resReg;
+   }
+
+TR::Register *OMR::Power::TreeEvaluator::vreductionAddWordHelper(TR::Node *node, TR::CodeGenerator *cg, TR::DataType type)
+   {
+   TR::Node *firstChild = node->getFirstChild();
+
+   TR::Register *srcReg = cg->evaluate(firstChild);
+   TR::Register *resReg;
+
+   TR::Register *shiftReg = cg->allocateRegister(TR_VRF);
+   TR::Register *temp = cg->allocateRegister(TR_VRF);
+   TR::Register *tempGPR;
+   TR::Register *tempVSX;
+
+   //choose addOp and allocate registers based on data type
+   TR::InstOpCode::Mnemonic addOp;
+
+   if (type == TR::Int32)
+   {
+      addOp = TR::InstOpCode::vadduwm;
+      tempVSX = cg->allocateRegister(TR_VRF);
+      tempGPR = cg->allocateRegister();
+      resReg = tempGPR;
+   }
+   else if (type == TR::Float)
+   {
+      addOp = TR::InstOpCode::xvaddsp;
+      tempVSX = cg->allocateRegister(TR_FPR);
+      resReg = tempVSX;
+   }
+   else
+   {
+      TR_ASSERT(false, "cannot call vreductionAddWordHelper on vector type %s\n", type.toString());
+      return NULL;
+   }
+
+   node->setRegister(resReg);
+
+   //set shift amount to -32 bits. Since only the lowest 6 bits of shiftReg are used by vrld as an unsigned integer, this will result in a shift of 32 bits
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisw, node, shiftReg, -16);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vadduwm, node, shiftReg, shiftReg, shiftReg);
+
+   //Evaluate sum across vector operand by rotating and performing lanewise addition
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vrld, node, temp, srcReg, shiftReg);
+   generateTrg1Src2Instruction(cg, addOp, node, tempVSX, srcReg, temp);
+   generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::xxpermdi, node, temp, temp, temp, 2);
+   generateTrg1Src2Instruction(cg, addOp, node, tempVSX, tempVSX, temp);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vrld, node, temp, temp, shiftReg);
+   generateTrg1Src2Instruction(cg, addOp, node, tempVSX, tempVSX, temp);
+
+   //Copy result into GPR if data type is Integer, convert to double precision format if data type is Float
+   if (type == TR::Int32)
+   {
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrd, node, tempGPR, tempVSX);
+      cg->stopUsingRegister(tempVSX);
+   }
+   else // type === TR::Float
+   {
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::xscvspdpn, node, tempVSX, tempVSX);
```

Comment:
It looks like this returns a double to me? I think the result here needs to be converted to a float.

Reply:
The Power ISA states the following: 
![image](https://user-images.githubusercontent.com/43620963/183963294-26023703-149c-4d3e-ae18-733efc67f301.png)
So I believe the compiler interprets it as a single-precision floating point value, even though it's in double-precision format

Reply:
Okay, that sounds good to me.